README.md
# Analysis of Public High Schools Admitting Students by Examination in Turkey


This repo contains a flexible, easy to understand and modifiable foundation for scraping, cleaning, analyzing and visualizing **publicly available education data of Turkey.**

Code was written in Python 3.8.5

Along with the foundation, different studies can be done about high schools and universities in Turkey. We conducted a study on the efficiency of public high schools in Turkey, as detailed in the following papers:


**[Analysis of Public High Schools Admitting Students by Examination in Turkey](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3939070)**,
*Beyza Arslan, Anıl Şen.*

***This study has received second prize in the Education Category from The Scientific and Technological Research Council of Turkey.***

If you use this code or data in your research, please cite us using this BibTeX entry:

```
@article{arslan2021analysis,
  title={Analysis of Public High Schools Admitting Students by Examination in Turkey},
  author={Arslan, Beyza and {\c{S}}en, An{\i}l},
  journal={Available at SSRN 3939070},
  year={2021}
}
```

## Contact Us

If you're interested in extending this work, have an idea or any questions:
- email us  asen16@ku.edu.tr or barslan16@ku.edu.tr 

or submit an issue.

## Installation Instructions

To get started, you'll need to have Python 3.8+ installed.


### Installing from Source

1. Clone this repository to your local machine:

  ```
   git clone https://github.com/asen16/high-schools-analysis.git
   ```

2. Change directory to the directory where requirements.txt is located.


  ```pyfunctiontypecomment
   cd [Path]
   ```

3. Run:

 ```pyfunctiontypecomment
   pip install -r requirements.txt
   ```
   in your shell.


Installation of Web Driver: You'll need to install the Web Driver to scrape data. You can follow the necessary installation steps from Selenium's documentation.

Selenium:
- See [Web Driver Installation](https://www.selenium.dev/selenium/docs/api/py/index.html#installing)


## Getting Started

### Paths
You can follow the instructions in these files to work on the analysis or data scraping part. Google colab version of the codes will be uploaded in next days.

- [`analysis`](https://github.com/asen16/high-schools-analysis/tree/main/analysis) : Shows how master data is filtered, analyzed and data visualized.
- [`data_scraping`](https://github.com/asen16/high-schools-analysis/tree/main/data_scraping): Explains how raw data is scraped from its source.

## Structure of the Code

The code repository is organized into the following components:

- The datasets are located in the `analysis/data` folder.
- The graphs are located in the `analysis/graphs` folder.
- The tabels are located in the `analysis/table` folder.
- The visualization and analysis programs are located in the `analysis` folder.
- The raw data are located in the `data_scraping/DATA` folder.
- The data scraping program is located in the `data_scraping` folder.

## Releases and Contributing

- Please let us know if you encounter any bugs by filing a Github issue.
- We appreciate all your contributions. If you plan to contribute a new Method, Data, or anything else, please see our [contribution guidelines](https://www.github.com/asen16/high-schools-analysis/blob/main/CONTRIBUTING.md).

## Changelog

For the complete release history, see [CHANGELOG.md](https://www.github.com/asen16/high-schools-analysis/blob/main/CHANGELOG.md).

## License

Analysis of Public High Schools Admitting Students by Examination in Turkey is released under the [MIT License](https://www.github.com/asen16/high-schools-analysis/blob/main/LICENSE).


CHANGELOG.md
# Changelog

# Release 1.0 (2021-11-07)

## Tutorials and Documentation

- Analysis and data visualization files.


analysis/bubble_graphs.py
# -*- coding: utf-8 -*-
"""
Created on Thu Nov  4 01:13:01 2021

@author: Anil Sen & Beyza Arslan

"""

#call required elements from graph_design
from graph_design import font, color_discrete_sequence_bubble, title_font_size
#call data from data_filtering
from data_filtering import highschool, provinces

#visualization library
import plotly.express as px
import json
import pandas as pd
import numpy as np


#It is the file where the graphics will be created.
graph_location="graphs/"


#===============================================================================#

#Average Score of High Schools According to the 2019 University Entrance Exam
#vs
#Percentile of High Schools According to the 2019 High School Entrance Exam
#High School Factor Bubble Graph

#===============================================================================#

fig = px.scatter(highschool,
                 x="lowest_score", 
                 y="percentile_of_2019", 
                 color="high_school_type",
                 size="high_school_quota_2019",
                 hover_name="high_school_name",
                 color_discrete_sequence=color_discrete_sequence_bubble,
                 #set max bubble size
                 size_max=20,
                 log_y=True,
                 #custom data for hover template
                 custom_data=['high_school_name', 'high_school_quota_2019', 'high_school_type'])

fig.update_layout(
    xaxis_title="Average Score of High Schools According to the 2019 University Entrance Exam",
    yaxis_title="Percentile of High Schools According to the 2019 High School Entrance Exam",
    legend_title=" High School Type",
        
    legend=dict(
        yanchor="bottom",
        y=0.01,
        xanchor="left",
        x=0.01
    ),
    font=dict(
        family=font,  
        size=8,
        color="#011126"
    )
)

#html type: <br> = space ; <b> %{data} </b> = bold ;
#To add data to hover template,  custom data list must be added to px.scatter
fig.update_traces(
    hovertemplate="<br>".join([
        "<b> %{customdata[0]} </b>",
        "<br>"
        "Average University Entrance Exam Score: %{x}",
        "High School Percentile: %{y}",
        "High School Type: %{customdata[2]}",
        "High School Quota: %{customdata[1]}",
    ])
)


fig.update_xaxes(title_font_size=title_font_size)
fig.update_yaxes(title_font_size=title_font_size)

#don't show color axes 
fig.update_coloraxes(showscale=False)

#if you are in spyder, check top right section and click Plots, you can see the graph!
fig.show()

#create html type graph
fig.write_html(graph_location+'high_school_exam_bubble.html', auto_open=auto_open)
#create pdf type graph
fig.write_image(graph_location+"high_school_exam_bubble.pdf")























analysis/data_filtering.py
# -*- coding: utf-8 -*-
"""

@author: Anil Sen & Beyza Arslan

"""

import pandas as pd
import numpy as np

#cleaned and matched ready master dataset
high_schools_data="data/public_high_schools_data.xlsx"

#GDP per capita data for each province
gdp_file='data/gdp_data.xlsx'

#read master data
data = pd.read_excel(high_schools_data, "Data")

#read gdp per capita data
gdppercapita = pd.read_excel(gdp_file, " Per capita GDP ($)")

#gerekli mi emin değilim
#gdppercapita = gdppercapita.apply(lambda x: x.astype(str).str.upper())



# add gdp per capita data to master data
gdpdata=[]
for province in data["province"]:
    gdpdata.append(round(float(gdppercapita[2019][pd.Index(gdppercapita["province"]).get_loc(province)]),2))

data["gdpdata"]=gdpdata

#"Pansiyon(Kız/Erkek)","Pansiyon(Erkek)" and "Pansiyon(Kız)" mean that the high school has a dormitory.
#"Pansiyon yok" means that the high school has not a dormitory.

#Instead of these values, we wrote 1 if the high school has a dormitory and 0 if it does not.
data["high_school_dormitory"] = data["high_school_dormitory"].map({"Pansiyon(Kız/Erkek)":1, "Pansiyon(Erkek)" : 1, "Pansiyon(Kız)" : 1, "Pansiyon Yok" : 0})

#""Hazırlık + 4 yıl"" means that the high school has a prep school.
#"4 yıl" means that the high school has not a prep school.

#Instead of these values, we wrote 1 if the high school has a prep school and 0 if it does not.
data["high_school_with_prep"] = data["high_school_with_prep"].map({"Hazırlık + 4 yıl":1, "4 yıl" : 0})

university_exam_data=data

#If you want, you can work on this excel file by 
#converting this dataframe to an excel file.

#university_exam_data.to_excel("data/university_exam_dataset.xlsx",sheet_name='Data')


#you can run the following loop to create the lists
"""
for i in range(1, 7):
    command_variable = data + str(i) + " = [] "
    exec(command_variable)
"""

#veride öğrencilerin bilgileri yer alıyor, liselerin bilgilerinin yer aldığı bir dataframe oluşturacağız.
data1,data2,data3,data4,data5,data6=[],[],[],[],[],[]

datahschool1,datahschool2,datahschool3,datahschool4=[],[],[],[]
datahschool5,datahschool6,datahschool7,datahschool8=[],[],[],[]




# tüm unique lise isimlerini loop'a alıyoruz
for i in pd.unique(data["high_school_name"]):
    #bu lisede yer alan tüm öğrencilerin verisi
    k_data=data.loc[data['high_school_name']==i]
    #okulun öğrenci sayısının birden fazla olması gerekiyor
    if len(k_data)>0:
        #yeni mezun olan öğrencilerin verilerini filtreliyor
        newly=k_data[k_data["newly_graduated_student"].values==1]
        #eğer yeni öğrenci sayısı birden fazlaysa,
        #bu öğrencilerin diploma notlarının ve üniversite sınavı
        #puanlarının ortalamasını alıp listelere ekliyoruz
        if len(newly) >0:
           data1.append(np.average(newly["lowest_score"]))
           data4.append(np.average(newly["average_diploma_grade"]))
        #hiç öğrenci yoksa 0'lar ekliyoruz.
        else:
            data1.append(0)
            data4.append(0)
        #önceden mezun olan öğrencilerin verilerini filtreliyor
        grad=k_data[k_data["former_graduate_student"].values==1]
        #eğer önceden mezun olan öğrenci sayısı birden fazlaysa,
        #bu öğrencilerin diploma notlarının ve üniversite sınavı
        #puanlarının ortalamasını alıp listelere ekliyoruz
        if len(grad) >0:
            data5.append(np.average(grad["lowest_score"]))
            data6.append(np.average(grad["average_diploma_grade"]))
        #hiç öğrenci yoksa 0'lar ekliyoruz.
        else:
            data5.append(0)
            data6.append(0)
        
        #sırayla gerekli bilgileri listelere ekliyoruz.
            
        #lisenin toplam yeni mezun ettiği öğrenci sayısı
        data2.append(np.sum(k_data["newly_graduated_student"]))
        #lisenin toplam eskiden mezun ettiği öğrenci sayısı
        data3.append(np.sum(k_data["former_graduate_student"]))
        #lisenin ismi
        datahschool1.append(np.array(k_data["high_school_name"])[0])
        #lisenin percentile'ı
        datahschool2.append(np.array(k_data["percentile_of_2019"])[0])
        #lisenin kotası
        datahschool3.append(np.array(k_data["high_school_quota_2019"])[0])
        #lisenin prep school'unun olup olmamaması
        datahschool4.append(np.array(k_data["high_school_with_prep"])[0])
        #lisenin yurdunun olup olmamaması
        datahschool5.append(np.array(k_data["high_school_dormitory"])[0])
        #lisenin türü
        datahschool6.append(np.array(k_data["high_school_type"])[0])
        #lisenin bulunduğu ilin gdp per capitası
        datahschool7.append(np.array(k_data["gdpdata"])[0])
        #lisenin bulunduğu il
        datahschool8.append(np.array(k_data["province"])[0])

#oluşturduğumuz listelerle bir dataframe oluşturuyoruz.    

highschool=pd.DataFrame({"lowest_score":data1,
                     "newly_graduated_student":data2,
                     "former_graduate_student":data3,
                     "average_diploma_grade":data4,
                     "high_school_name":datahschool1,
                     "percentile_of_2019":datahschool2,
                     "high_school_quota_2019":datahschool3,
                     "high_school_with_prep":datahschool4,
                     "high_school_dormitory":datahschool5,
                     "high_school_type":datahschool6,
                     "gdpdata":datahschool7,
                     "province":datahschool8}           
    )

#liselerin faktör verimliliklerini aşağıdaki formülle hesaplıyoruz:
    #(newly graduated student* high school average university exam score)/high school quota

highschool["factor"]=list((highschool["newly_graduated_student"]*highschool["lowest_score"])/highschool["high_school_quota_2019"])

#differanibale olabilmesi için boyutlar arasındaki farkı arttıyoruz.
highschool["bubblesize"]=(highschool["high_school_quota_2019"]**1.5)/10

#her lise tipinin açıklamasını ekliyoruz
high_school_type=[]
for number in highschool["high_school_type"]:
    if number==1:
        high_school_type.append("Science High School")
    
    if number==2:
        high_school_type.append("Anatolian High School")

    if number==3:
        high_school_type.append("Anatolian Imam Hatip High School")

    if number==4:
        high_school_type.append("Vocational High School")

    if number==5:
        high_school_type.append("Social Science High School")

#yukarıdaki veriyi dataframe'e ekliyoruz.
highschool["high_school_type"]=high_school_type

#If you want, you can work on this excel file by 
#converting this dataframe to an excel file.

#highschool.to_excel("data/high_school_dataset.xlsx",sheet_name='Data')



#veride öğrencilerin bilgileri yer alıyor, şehirlerin bilgilerinin yer aldığı bir dataframe oluşturacağız.
#yukarıda süreci şehirler için tekrar ediyoruz.

province1,province2,province3,province4,province5,province6=[],[],[],[],[],[]

provincedata1,provincedata2,provincedata3,provincedata4=[],[],[],[]
provincedata5,provincedata6,provincedata7,provincedata8=[],[],[],[]


for i in pd.unique(data["province"]):
    k_data=data.loc[data['province']==i]
    if len(k_data)>0: 
        newly=k_data[k_data["newly_graduated_student"].values==1]
        if len(newly) >0:
           province1.append(np.average(newly["lowest_score"]))
           province4.append(np.average(newly["average_diploma_grade"]))
        else:
            province1.append(0)
            province4.append(0)
        grad=k_data[k_data["former_graduate_student"].values==1]
        if len(grad) >0:
            province5.append(np.average(grad["lowest_score"]))
            province6.append(np.average(grad["average_diploma_grade"]))
        else:
            province5.append(0)
            province6.append(0)
        province2.append(np.sum(k_data["newly_graduated_student"]))
        province3.append(np.sum(k_data["former_graduate_student"]))
        provincedata2.append(np.average(k_data["percentile_of_2019"]))
        holding="0"
        k_quota=0
        for i in range(len(k_data["high_school_name"])):
            if np.array(k_data["high_school_name"])[i]==holding:
                pass
            else:
                k_quota+=np.array(k_data["high_school_quota_2019"])[i]
                holding=np.array(k_data["high_school_name"])[i]
        provincedata3.append(k_quota)
        k_quota=0
        provincedata4.append(np.average(k_data["high_school_with_prep"]))
        provincedata5.append(np.average(k_data["high_school_dormitory"]))
        provincedata7.append(np.array(k_data["gdpdata"])[0])
        provincedata8.append(np.array(k_data["province"])[0])


provinces=pd.DataFrame()

provinces=pd.DataFrame({"lowest_score":province1,
                     "newly_graduated_student":province2,
                     "former_graduate_student":province3,
                     "average_diploma_grade":province4,
                     "percentile_of_2019":provincedata2,
                     "high_school_quota_2019":provincedata3,
                     "high_school_with_prep":provincedata4,
                     "high_school_dormitory":provincedata5,
                     "gdpdata":provincedata7,
                     "province":provincedata8}           
    )


provinces["factor"]=list((provinces["newly_graduated_student"]*provinces["lowest_score"])/provinces["high_school_quota_2019"])


#If you want, you can work on this excel file by 
#converting this dataframe to an excel file.

#provinces.to_excel("data/provinces_dataset.xlsx",sheet_name='Data')


analysis/graph_high_school_exam_bubble.py
# -*- coding: utf-8 -*-
"""
Created on Thu Nov  4 01:13:01 2021

@author: Anil Sen & Beyza Arslan

"""

#call required elements from graph_design
from graph_design import font, color_discrete_sequence_bubble, title_font_size,auto_open
#call data from data_filtering
from data_filtering import highschool
#visualization library
import plotly.express as px


#It is the file where the graphics will be created.
graph_location="graphs/"


#===============================================================================

#Average Score of High Schools According to the 2019 University Entrance Exam
#vs
#Percentile of High Schools According to the 2019 High School Entrance Exam
#High School Exam Bubble Graph

#===============================================================================

fig = px.scatter(highschool,
                 x="lowest_score", 
                 y="percentile_of_2019", 
                 color="high_school_type",
                 size="high_school_quota_2019",
                 hover_name="high_school_name",
                 color_discrete_sequence=color_discrete_sequence_bubble,
                 #set max bubble size
                 size_max=20,
                 log_y=True,
                 log_x=True,
                 #custom data for hover template
                 custom_data=['high_school_name', 'high_school_quota_2019', 'high_school_type'])

fig.update_layout(
    xaxis_title="Average Score of High Schools According to the 2019 University Entrance Exam",
    yaxis_title="Percentile of High Schools According to the 2019 High School Entrance Exam",
    legend_title=" High School Type",
        
    legend=dict(
        yanchor="bottom",
        y=0.01,
        xanchor="left",
        x=0.01
    ),
    font=dict(
        family=font,  
        size=8,
        color="#011126"
    )
)

#html type: <br> = space ; <b> %{data} </b> = bold ;
#to add data to hover template, px.scatter custom data list must be added
fig.update_traces(
    hovertemplate="<br>".join([
        "<b> %{customdata[0]} </b>",
        "<br>"
        "Average University Entrance Exam Score: %{x}",
        "High School Percentile: %{y}",
        "High School Type: %{customdata[2]}",
        "High School Quota: %{customdata[1]}",
    ])
)

#set font size
fig.update_xaxes(title_font_size=title_font_size)
fig.update_yaxes(title_font_size=title_font_size)

#dont show color axes
fig.update_coloraxes(showscale=False)

#if you are in spyder, check top right section and click Plots, you can see the graph!
fig.show()

#create html type graph
fig.write_html(graph_location+'graph_high_school_exam_bubble.html', auto_open=auto_open)
#create pdf type graph
fig.write_image(graph_location+"graph_high_school_exam_bubble.pdf")


analysis/graph_province_efficiency_factor_bubble.py
# -*- coding: utf-8 -*-
"""
@author: Anil Sen & Beyza Arslan
"""

#call required elements from graph_design
from graph_design import font, title_font_size,auto_open
#call data from data_filtering
from data_filtering import provinces

#visualization library
import plotly.express as px
import pandas as pd


#It is the file where the graphics will be created.
graph_location="graphs/"

#===============================================================================#

#Average Efficiency Factors of High Schools
#vs
# Average Percentile of High Schools <br>According to High School Entrance Exam
#Province Factor Bubble Graph

#===============================================================================#

#round to two decimals
provinces["factor"]=[round(x,2) for x in provinces["factor"]]    
provinces["lowest_score"]=[round(x,2) for x in provinces["lowest_score"]]    

#use mono color
colors=[]
for i in range(len(provinces["factor"])):
    colors.append(1)

fig = px.scatter(provinces,
                 x="factor",
                 y="percentile_of_2019",
                 text="province",
                 size="high_school_quota_2019",
                 hover_name="province",
                 color=colors, 
                 color_continuous_scale="blues",  
                 range_color=(0,1),
                 log_y=True,
                 custom_data=['factor', 'lowest_score',"high_school_quota_2019","province"]
                 )


#html type: <br> = space ; <b> %{data} </b> = bold ;
#To add data to hover template,  custom data list must be added to px.scatter
fig.update_traces(
    textposition='top center',
    hovertemplate="<br>".join([
        "<b> %{customdata[3]} </b>",
        "<br>"
        "Province Efficiency Factor: %{customdata[0]}",
        "Province Exam Score: %{customdata[1]}",
        "Province High School Quota: %{customdata[2]}",

    ])
)


fig.update_layout(

    xaxis_title=" Average Efficiency Factors of High Schools",
    yaxis_title=" Average Percentile of High Schools <br>According to High School Entrance Exam",

    font=dict(
        family=font,
        size=5.5,
        color="#011126"
    ),
    legend=dict(
        yanchor="top",
        y=0.99,
        xanchor="right",
        x=1.00
    ),
    #background color
    paper_bgcolor="#FFFFFF",
    yaxis = dict(tickfont = dict(size=10)),
    xaxis = dict(tickfont = dict(size=10))
    
)


fig.update_xaxes(title_font_size=title_font_size)
fig.update_yaxes(title_font_size=title_font_size)

#don't show color axes 
fig.update_coloraxes(
    showscale=False,
    #colorbar_tickfont_size=22,
)


#if you are in spyder, check top right section and click Plots, you can see the graph!
fig.show()


#create html type graph
fig.write_image(graph_location+"graph_province_efficiency_factor_bubble.pdf")
#create pdf type graph
fig.write_html(graph_location+'graph_province_efficiency_factor_bubble.html', auto_open=auto_open)

analysis/graph_high_school_efficiency_factor_bubble.py
# -*- coding: utf-8 -*-
"""
Created on Sat Nov  6 21:20:47 2021

@author: anils
"""


#call required elements from graph_design
from graph_design import font, color_discrete_sequence_bubble, title_font_size,auto_open
#call data from data_filtering
from data_filtering import highschool
#visualization library
import plotly.express as px


#It is the file where the graphics will be created.
graph_location="graphs/"


# =============================================================================
#High Schools' Efficiency Factors
#vs
#Percentile of High Schools According to the 2019 High School Entrance Exam
#High School Efficiency Factor Bubble Graph
# =============================================================================


fig = px.scatter(highschool,
                 x="factor", 
                 y="percentile_of_2019", 
                 color="high_school_type",
                 size="high_school_quota_2019",
                 #hover_name="high_school_name",
                 color_discrete_sequence=color_discrete_sequence_bubble,
                 #set max bubble size
                 size_max=20,
                 log_y=True,
                 #custom data for hover template
                 custom_data=['high_school_name', 'high_school_quota_2019', 'high_school_type'])

fig.update_layout(
    
    xaxis_title="High Schools' Efficiency Factors",
    yaxis_title="Percentile of High Schools According to the 2019 High School Entrance Exam",
    legend_title=" High School Type",
    
    legend=dict(
        yanchor="top",
        y=0.99,
        xanchor="right",
        x=1.00
    ),
    font=dict(
        family=font,  
        size=8,
        color="#011126"
    )
)
#html type: <br> = space ; <b> %{data} </b> = bold ;
#to add data to hover template, px.scatter custom data list must be added
fig.update_traces(
    hovertemplate="<br>".join([
        "<b> %{customdata[0]} </b>",
        "<br>"
        "High School Efficiency Factor: %{x}",
        "High School Percentile: %{y}",
        "High School Type: %{customdata[2]}",
        "High School Quota: %{customdata[1]}",
    ])
)


fig.update_xaxes(title_font_size=title_font_size)
fig.update_yaxes(title_font_size=title_font_size)

#dont't show color axes
fig.update_coloraxes(showscale=False)

#if you are in spyder, check top right section and click Plots, you can see the graph!
fig.show()


#create html type graph
fig.write_html(graph_location+'graph_high_school_efficiency_factor_bubble.html', auto_open=auto_open)
#create pdf type graph
fig.write_image(graph_location+"graph_high_school_efficiency_factor_bubble.pdf")









analysis/regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon May  3 15:28:32 2021

@author: Anil Sen & Beyza Arslan
"""

# =============================================================================
# Importing necessary libraries
# =============================================================================

import pandas as pd
import numpy as np

#for dataset
from data_filtering import highschool
from data_filtering import university_exam_data

#for regression/statistical analysis
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import variance_inflation_factor
import statsmodels.api as sm
from patsy import dmatrices

#for visualization
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

#for LaTex table
from stargazer.stargazer import Stargazer
from IPython.core.display import HTML

# =============================================================================
# =============================================================================
# EFFICIENCY FACTOR REGRESSION
# =============================================================================
# =============================================================================

#define location for output
graph_location="graphs/"
table_location="table/"


# =============================================================================
# Checking correlation and visualize
# =============================================================================

#create new dataframe and remove unnecessary columns
to_corr_check = highschool
to_corr_check = to_corr_check.drop(columns = ["factor", "bubblesize", "province", "lowest_score" ])

#find correlation coefficients
to_corr_check.corr()

#visualize
sns.heatmap(to_corr_check.corr())

#save as png file
plt.savefig(graph_location+'correlation-coefficients.png')

# =============================================================================
# Creating OLS Model for Efficiency Factor Regression
# =============================================================================

#remove 0 values of factor
highschool= highschool[highschool['factor'] != 0]

#insert log(percentile_of_2019) to dataframe
highschool.insert(5, "log_percentile_of_2019", np.log(highschool["percentile_of_2019"]))

#create OLS Model - Define Y and X variables
factor_model = ols('factor ~   log_percentile_of_2019  + C(high_school_type)  +   average_diploma_grade + high_school_with_prep + high_school_dormitory + gdpdata ', data=highschool)
y = highschool["factor"]
X = highschool[['log_percentile_of_2019','high_school_type' , "average_diploma_grade" , "high_school_with_prep" , "high_school_dormitory" , "gdpdata"]]

#fit Model
fitted_factor_model = factor_model.fit()
fitted_factor_model.get_robustcov_results().summary()


# =============================================================================
# Find Predicted Values
# =============================================================================

X = sm.add_constant(X)
predicted_values = fitted_factor_model.predict(X)

#visualize predicted vs real values
highschool = highschool.rename_axis('index1').reset_index()
highschool.insert(1, "predicted_factor", predicted_values)
highschool.insert(2, 'real_factor', highschool['factor'])
fig_prediction_vs_factor = px.line(highschool, x="index1", y=highschool.columns[1:3] )
fig_prediction_vs_factor.update_traces(mode="lines", hovertemplate=None)
fig_prediction_vs_factor.update_layout(hovermode="x")
fig_prediction_vs_factor.update_layout(
    xaxis_title="Indexes",
    yaxis_title="Predicted and Real Factor Value ",
    legend_title="Predicted vs Real",
        
    legend=dict(
        yanchor="bottom",
        y=0.01,
        xanchor="left",
        x=0.01
    ),
    font=dict(
        family="Roboto",  
        size=8,
        color="#011126"
    )
)
fig_prediction_vs_factor.show()
fig_prediction_vs_factor.write_html(graph_location+"prediction_vs_factor.html",auto_open=True)

# =============================================================================
# Hypothesis Testing and Check Multicollinearity
# =============================================================================

#find t-score
hypothesis = 'high_school_dormitory = 0' #can used with other hypothesis
t_test = fitted_factor_model.t_test(hypothesis)

#calculate VIF 
y, X = dmatrices('factor ~   log_percentile_of_2019  + high_school_type  +   average_diploma_grade + high_school_with_prep + high_school_dormitory + gdpdata ', data=highschool, return_type='dataframe')
vif_factor = pd.DataFrame()
vif_factor['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
vif_factor['variable'] = X.columns

# =============================================================================
# Create LaTex Table
# =============================================================================

stargazer = Stargazer([fitted_factor_model])
HTML(stargazer.render_html())
f = open(table_location+'my_factor_reg.tex', 'w')
f.write(stargazer.render_latex())
f.close()

# =============================================================================
# =============================================================================
# EXAM SCORE REGRESSION
# =============================================================================
# =============================================================================


# =============================================================================
# Creating OLS Model for Exam Score Regression
# =============================================================================

#remove 0 values of factor
university_exam_data= university_exam_data[university_exam_data['lowest_score'] != 0]

#insert log(percentile_of_2019) to dataframe
university_exam_data.insert(5, "log_percentile_of_2019", np.log(university_exam_data["percentile_of_2019"]))

#create OLS Model - Define Y and X variables
exam_model = ols('lowest_score ~   log_percentile_of_2019  + C(high_school_type)  + newly_graduated_student +  average_diploma_grade + high_school_with_prep + high_school_dormitory + gdpdata ', data=university_exam_data)
y_exam = university_exam_data["lowest_score"]
X_exam = university_exam_data[['log_percentile_of_2019','high_school_type' ,"newly_graduated_student", "average_diploma_grade" , "high_school_with_prep" , "high_school_dormitory" , "gdpdata"]]

#fit Model
fitted_exam_model = exam_model.fit()
fitted_exam_model.get_robustcov_results().summary()


# =============================================================================
# Find Predicted Values
# =============================================================================

X_exam = sm.add_constant(X_exam)
predicted_values_exam = fitted_exam_model.predict(X_exam)

#visualize predicted vs real values
university_exam_data = university_exam_data.rename_axis('index1').reset_index()
university_exam_data.insert(1, "predicted_exam_score", predicted_values_exam)
university_exam_data.insert(2, 'real_score', university_exam_data['lowest_score'])
fig_prediction_vs_exam = px.line(university_exam_data, x="index1", y=university_exam_data.columns[1:3] )
fig_prediction_vs_exam.update_traces(mode="lines", hovertemplate=None)
fig_prediction_vs_exam.update_layout(hovermode="x")
fig_prediction_vs_exam.update_layout(
    xaxis_title="Indexes",
    yaxis_title="Predicted and Real Exam Score",
    legend_title="Predicted vs Real",
        
    legend=dict(
        yanchor="bottom",
        y=0.01,
        xanchor="left",
        x=0.01
    ),
    font=dict(
        family="Roboto",  
        size=8,
        color="#011126"
    )
)
fig_prediction_vs_exam.show()
fig_prediction_vs_exam.write_html(graph_location+"prediction_vs_exam.html",auto_open=True)

# =============================================================================
# Hypothesis Testing and Check Multicollinearity
# =============================================================================

#find t-score
hypothesis_exam = 'average_diploma_grade = 0' #can used with other hypothesis
t_test = fitted_exam_model.t_test(hypothesis_exam)

#calculate VIF 
y_exam, X_exam = dmatrices('lowest_score ~   log_percentile_of_2019  + C(high_school_type)  + newly_graduated_student +  average_diploma_grade + high_school_with_prep + high_school_dormitory + gdpdata ', data=university_exam_data, return_type='dataframe')
vif_exam = pd.DataFrame()
vif_exam['VIF'] = [variance_inflation_factor(X_exam.values, i) for i in range(X_exam.shape[1])]
vif_exam['variable'] = X_exam.columns

# =============================================================================
# Create LaTex Table
# =============================================================================

stargazer = Stargazer([fitted_exam_model])
HTML(stargazer.render_html())
f = open(table_location+'my_exam_reg.tex', 'w')
f.write(stargazer.render_latex())
f.close()



